### ISSUE TITLE
Wrong `Eager Loading Detected` if use `before_all` or `let_it_be`

### ISSUE THREAD

**Issue by chaomao** [flyerhzm/bullet#666]
Hi, zm

recently I start refactoring our test, move large object creation from `let` to [let_it_be](https://test-prof.evilmartians.io/#/recipes/let_it_be), which will change object creation from `before_each` hook into `before_all` hook, then I found some tests failed due to `eager loading detected`, like below
```ruby
RSpec.describe Post, type: :model do
  # post has_many comments
  # comment belongs_to post
  # creator belongs_to comment
  let_it_be(:post) { create(:post) }
  let_it_be(:comment) { create(:comment, post: post)}

  # before(:all) do # same as let_it_be
  #   puts "start before all"
  #   @post = create(:post)
  #   @comment = create(:comment, post: @post)
  #   puts "end before all"
  # end

  it 'ok' do
    creator = create(:creator, comment_id: comment.id)
    comment.post # ok
  end

  it 'raise N+1 query issue' do
    creator = create(:creator, comment: comment)
    comment.post # will raise eager loading Comment => [:post]
  end
end
```
which will raise error `USE eager loading detected, Comment => [:post], Add to your query: .includes([:post])`, reproducable repo is [here](https://github.com/chaomao/bullet_bug/blob/master/spec/models/post_spec.rb). I don't think the detection is correct :(

I read the source code of bullet and found the issue is the **object created in `before_all` hook didn't add into `Bullet::Detector::NPlusOneQuery::impossible_objects`**, because Bullet.start_request is invoked in `before_each` hook.
```ruby
# n_plus_one_query.rb
        def add_impossible_object(object)
          return unless Bullet.start? # will return when object creation in `before_all` hook
          return unless Bullet.n_plus_one_query_enable?
          return unless object.bullet_primary_key_value

          Bullet.debug('Detector::NPlusOneQuery#add_impossible_object', "object: #{object.bullet_key}")
          impossible_objects.add object.bullet_key
        end

# spec_helper.rb, will start Bullet in before_each hook.
if Bullet.enable?
  RSpec.configure do |config|
    config.before do
      Bullet.start_request
    end

    config.after do
      Bullet.perform_out_of_channel_notifications if Bullet.notification?
      Bullet.end_request
    end
  end
end
```

so my questions are:
1. is there a way to avoid `eager loading detected` error when I use `before_all` or `let_it_be`? I really want to move some large object creation out of each test, which will significantly improve performance
2. no matter the problem is solved or not, I hope the documentation can cover this scenario, it may help other people on it :)

**Comment by flyerhzm** [comment]
@chaomao The solution comes to my mind is

1. preserve all bullet objects, `possible_objects`, `impossible_objects`, etc., after `before_all`
2. restore all bullet objects before `before_each` or each test.
3. reset all bullet objects except preserved objects after each test or `after_each`.
4. reset all preserved bullet objects after `after_all`.

But it's not simple and I'm not sure if I have time to do it right now.

Or you can skip the bullet request for some tests which contains before_all or let_it_be.

**Comment by chaomao** [comment]
thanks for the reply, I will skip bullet request for some tests for now :)

**Comment by jamesmcguirepro** [comment]
I'm encountering this issue as well.

A workaround we discovered is to use `let_it_be_with_reload` to avoid the bullet warnings

**Comment by erimicel** [comment]
Yes `let_it_be(:record, reload: true) { create(:record) }` works
